core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1048,9 +1059,16 @@ class Log(@volatile var dir: File,
       lock synchronized {
         // remove the segments for lookups
         deletable.foreach(deleteSegment)
-        logStartOffset = math.max(logStartOffset, segments.firstEntry().getValue.baseOffset)
+        logStartOffset = math.max(logStartOffset, segments.firstEntry.getValue.baseOffset)
         leaderEpochCache.clearAndFlushEarliest(logStartOffset)
-        producerStateManager.evictUnretainedProducers(logStartOffset)
+        producerStateManager.truncateHead(logStartOffset)
+        updateFirstUnstableOffset()
+      }
+    } else {
+      lock synchronized {
+        // The log start offset may now point to the middle of a segment, so we need to truncate the producer
+        // state to ensure that non-retained producers are evicted.
+        producerStateManager.truncateHead(logStartOffset)
```

Comment:
logStartOffset is only going to change when some segments are deleted in deleteSegments() or when maybeIncrementLogStartOffset() is called. So, it seems that we don't need to do anything in the else clause since the other two cases are already covered.